### ISSUE TITLE
Add elegant support

### ISSUE THREAD

**Issue by cchall** [radiasoft/rsopt#20]
With completion of https://github.com/radiasoft/sirepo/issues/2864 can now try adding elegant importer.

- [x] Add importer (in 94093fb3c31b6db07011a74fbb405b8e6342b77a)
- [x] ~~Editor to write parameter/settings changes to imported PKDict (validate off of https://github.com/radiasoft/sirepo/blob/master/sirepo/package_data/static/json/elegant-schema.json)~~
       The elements/commands in the `SimData` model have all names filled with defaults if not given from import. This removes the need to do explicit validation when parameters/settings are updated before write.
- [x] Write simulation files at each new job (~~probably requires https://github.com/radiasoft/sirepo/issues/2865 to be finished~~  -Done) 
(partially implemented in 94093fb3c31b6db07011a74fbb405b8e6342b77a. completed in a2b418e5688cc5e2e55d31dc47ea72dcde95e10c)

~~Full cycle also requires completion of last item in #18~~ - done

- [x] Debug handoff error occurring in libEnsemble (This proved to be nontrivial and actually is showing up for an example on master, see #22)

**Comment by robnagler** [comment]
I think we should encapsulate the validation in sirepo.lib. We already have some validation, but could write more. We can add that to sirepo.lib.Exporter or to a separate validation phase. Would be good to detail.

**Comment by cchall** [comment]
For validation my plan is to allow the rsopt configuration files to change elegant settings that may not have been explicitly entered in an input file. So this is really validation of data being entered in an rsopt specified format against elegant (which is conveniently stored by Sirepo schema files).

**Comment by robnagler** [comment]
I need an example to understand example what you mean.

**Comment by robnagler** [comment]
The import/export is implemented for elegant (only). There are a couple of tests, but the [easiest one is here](https://github.com/radiasoft/sirepo/blob/9d3bd33f88a5d7be28abd2c54c2d771b4dbe33fc/tests/template/elegant_import_test.py#L43). The one that imports the TESSA example [is here](https://github.com/radiasoft/sirepo/blob/9d3bd33f88a5d7be28abd2c54c2d771b4dbe33fc/tests/lib_test.py#L16).

The basic structure is:

```py
import sirepo.lib
d = sirepo.lib.Importer('elegant').parse_file('some-elegant.ele')
d.models.bunch.n_particles_per_bunch = 12345
d.write_files('some-dest-dir')
```

`d` is a subclass `SimData` of `PKDict` with some more state (see below) about the inputs. `some-elegant.ele` and `some-dest-dir` can be a `py.path` or strings. 

The input files are parsed from where `some-elegant.ele` lives. Then,`write_files` will read from that same directory (state in `d`) to get the files to output to `some-dest-dir`.

You probably will need to call `copy.deepcopy` on d, depending on how you generate the values for the tests, especially if the generator is threaded.

The output file names match how sirepo munges them so we may need to talk about that. The importer rewrites all the names.

I think we should add methods to `SimData` that do what you want. Let's discuss.

**Comment by cchall** [comment]
Sorry, @robnagler , this got put on the back burner for a little while, but it's proving very useful already!
A few notes for usage not necessarily specific to rsopt:

- output files: Writing all the supporting files (e.g. PC.param, knsl42.liwake.sdds in the TESSA example) is a nice but should probably be optional. For rsopt usage libEnsemble can handle sim linking files that need to be included in new simulation run directories which seems preferable to doing post-simulation cleanup. Is there an easy way to get a list of the files that `SimData` writes?

For the validation issue:
- Looking at  `SimData` from one of the examples I now realize that all the default fields for command namelists and lattice elements are already populated so part of the validation I was planning is already handled.
-  I started making use of the Importer already for some other work and ran into the issue that values in 'SimData' are stored as strings. When importing the lattice into a `StructuredBeamline` for manipulation (https://github.com/radiasoft/rsbeams/blob/6618cbc1b9d1018f003198282303d8aef5b048a5/rsbeams/rslattice/templates/lattice_importer.py#L37) I currently just have to cast values on a case-by-case basis like in https://github.com/radiasoft/rsbeams/blob/6618cbc1b9d1018f003198282303d8aef5b048a5/rsbeams/pkcli/lattice.py#L24. Would it be difficult to have an option to use the Sirepo schema to just recast all the values at import?

**Comment by robnagler** [comment]
> Writing all the supporting files (e.g. PC.param, knsl42.liwake.sdds in the TESSA example) is a nice but should probably be optional.

I'm trying to think of this as a general solution to the problem so that the interface is simple. We can symlink instead of copy in write_files (and rename if you like). Would that satisfy the cleanup problem?

> Is there an easy way to get a list of the files that SimData writes?

Yes. See https://github.com/radiasoft/sirepo/issues/2947

> Would it be difficult to have an option to use the Sirepo schema to just recast all the values at import?

Easy peasy (maybe :) https://github.com/radiasoft/sirepo/issues/2946

**Comment by cchall** [comment]
> I'm trying to think of this as a general solution to the problem so that the interface is simple. We can symlink instead of copy in write_files (and rename if you like). Would that satisfy the cleanup problem?

I think symlinked files are probably preferable in the majority of cases so lets just go with that. A rename option probably isn't needed. Once radiasoft/sirepo#2947 is in place they can always be overwritten easily if really needed for a particular case anyway.

**Comment by robnagler** [comment]
One concern with symlinks is related to "overwritten easily": originals (source of symlink) could get overwritten if "cp" is used. I'll change it, just noting.

fyi, the "rename" was about the function name itself.

**Comment by cchall** [comment]
merged into develop
